File: 2025/02 February/25 Histogram Max Rectangular Area/main.py
# Link: https://www.geeksforgeeks.org/problems/maximum-rectangular-area-in-a-histogram-1587115620/1


class Solution:
    def getMaxArea(self, arr):
        n = len(arr)
        stack = []
        ans = 0

        for i in range(n + 1):
            while stack and (i == n or arr[stack[-1]] >= arr[i]):
                height = arr[stack.pop()]
                width = i if not stack else (i - stack[-1] - 1)
                ans = max(ans, height * width)
            stack.append(i)

        return ans


if __name__ == "__main__":
    obj = Solution()
    arr = [60, 20, 50, 40, 10, 50, 60]
    print(obj.getMaxArea(arr))

    arr = [3, 5, 1, 7, 5, 9]
    print(obj.getMaxArea(arr))

    arr = [3]
    print(obj.getMaxArea(arr))

File: 2025/02 February/25 Histogram Max Rectangular Area/main.cpp
// Link: https://www.geeksforgeeks.org/problems/maximum-rectangular-area-in-a-histogram-1587115620/1

#include <iostream>
#include <vector>
#include <stack>
using namespace std;

class Solution
{
public:
    int getMaxArea(vector<int> &arr)
    {
        int n = arr.size();
        vector<int> prev(n, -1);
        vector<int> nex(n, n);
        stack<int> st;
        for (int i = 0; i < n; i++)
        {
            while (!st.empty() and arr[st.top()] >= arr[i])
            {
                st.pop();
            }
            if (!st.empty())
            {
                prev[i] = st.top();
            }
            st.push(i);
        }
        while (!st.empty())
        {
            st.pop();
        }
        for (int i = n - 1; i >= 0; i--)
        {
            while (!st.empty() and arr[st.top()] >= arr[i])
            {
                st.pop();
            }
            if (!st.empty())
            {
                nex[i] = st.top();
            }
            st.push(i);
        }
        int ans = 0;
        for (int i = 0; i < n; i++)
        {
            ans = max(ans, arr[i] * (nex[i] - prev[i] - 1));
        }
        return ans;
    }
};

int main()
{
    Solution object;
    vector<int> arr = {60, 20, 50, 40, 10, 50, 60};
    cout << object.getMaxArea(arr) << endl;

    arr = {3, 5, 1, 7, 5, 9};
    cout << object.getMaxArea(arr) << endl;

    arr = {3};
    cout << object.getMaxArea(arr) << endl;
    return 0;
}